### PR TITLE
Add voters for content elements

### PR DESCRIPTION
**Pull request by aschempp**

### CODE REVIEW

File: core-bundle/contao/drivers/DC_Table.php [hunk 1]
```diff
@@ -392,6 +392,8 @@ public function showAll()
 			);
 
 			$objSession->set('CLIPBOARD', $arrClipboard);
+
+			$this->redirect($this->getReferer());
```
Comment (leofeyer) [inline core-bundle/contao/drivers/DC_Table.php]: This needs to be reverted: https://github.com/contao/contao/pull/7055#pullrequestreview-1988503023

File: calendar-bundle/src/Security/Voter/CalendarContentVoter.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        return 'tl_content';
+    }
+
+    protected function hasAccessToRecord(TokenInterface $token, string $table, int $id): bool
+    {
+        if ('tl_calendar_events' !== $table) {
+            return true;
+        }
+
+        if (!$this->accessDecisionManager->decide($token, [ContaoCalendarPermissions::USER_CAN_ACCESS_MODULE])) {
```
Comment (leofeyer) [inline calendar-bundle/src/Security/Voter/CalendarContentVoter.php:53]:
```suggestion
        if (!$this->accessDecisionManager->decide($token, [ContaoCalendarPermissions::USER_CAN_ACCESS_MODULE], 'calendar')) {
```

Don‘t we need a subject here?

File: calendar-bundle/tests/Security/Voter/CalendarContentVoterTest.php [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+        $voter = new CalendarContentVoter($accessDecisionManager, $connection);
+        $decision = $voter->vote($token, $action, [ContaoCorePermissions::DC_PREFIX.'tl_content']);
+
+        $this->assertSame(VoterInterface::ACCESS_ABSTAIN, $decision);
+    }
+
+    public static function checksElementAccessPermissionProvider(): iterable
+    {
+        yield 'Check access to news archive when creating element in news' => [
```
Comment (leofeyer) [inline calendar-bundle/tests/Security/Voter/CalendarContentVoterTest.php]:
```suggestion
        yield 'Check access to calendar when creating element in events' => [
```

Needs to be changed in the lines below, too.

File: news-bundle/src/Security/Voter/NewsContentVoter.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        return 'tl_content';
+    }
+
+    protected function hasAccessToRecord(TokenInterface $token, string $table, int $id): bool
+    {
+        if ('tl_news' !== $table) {
+            return true;
+        }
+
+        if (!$this->accessDecisionManager->decide($token, [ContaoNewsPermissions::USER_CAN_ACCESS_MODULE])) {
```
Comment (leofeyer) [inline news-bundle/src/Security/Voter/NewsContentVoter.php:53]:
```suggestion
        if (!$this->accessDecisionManager->decide($token, [ContaoNewsPermissions::USER_CAN_ACCESS_MODULE], 'archive')) {
```

No?

**Review by leofeyer — COMMENTED** [review]
Changes look good 👍 

But tests are missing for the `ContentElementNestingVoter`. And there are quite a few uncovered lines in the `AbstractDynamicPtableVoter` that we should test as well.

File: calendar-bundle/src/Security/Voter/CalendarContentVoter.php (hunk 2 above)
Comment (aschempp) [reply calendar-bundle/src/Security/Voter/CalendarContentVoter.php:53]: No, `ContaoCalendarPermissions::USER_CAN_ACCESS_MODULE` already includes the subject 😎

File: news-bundle/src/Security/Voter/NewsContentVoter.php (hunk 4 above)
Comment (aschempp) [reply news-bundle/src/Security/Voter/NewsContentVoter.php:53]: no 🤪

File: core-bundle/src/EventListener/DataContainer/DefaultOperationsListener.php [hunk 5]
```diff
@@ -106,7 +106,7 @@ private function getDefaults(string $table): array
             if (DataContainer::MODE_TREE_EXTENDED !== ($GLOBALS['TL_DCA'][$ctable]['list']['sorting']['mode'] ?? null)) {
                 $operations += [
                     'children' => [
-                        'href' => 'table='.$ctable,
+                        'href' => 'table='.$ctable.($ctable === $table ? '&amp;ptable='.$table : ''),
```
Comment (leofeyer) [inline core-bundle/src/EventListener/DataContainer/DefaultOperationsListener.php:109]:
```suggestion
                        'href' => 'table='.$ctable.($hasPtable && $ctable === $table ? '&amp;ptable='.$table : ''),
```

Don‘t we have to check `$hasPtable` here as well?

Comment (aschempp) [reply core-bundle/src/EventListener/DataContainer/DefaultOperationsListener.php:109]: Not sure. This is the children operation, so in theory it is independent of the `ptable`. But if the current table and the **child table** is the same, we need to add `ptable=...` to the URL. That currently only applies to nested elements, but I guess you know that 😆

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.

File: core-bundle/src/EventListener/DataContainer/DefaultOperationsListener.php (hunk 5 above)
Comment (leofeyer) [reply core-bundle/src/EventListener/DataContainer/DefaultOperationsListener.php:109]: I ran some more tests and you are right. Using `$hasPtable` would be wrong here.

**Pull request merged**